required/latex-lab/latex-lab-math.dtx

```diff
@@ -685,8 +712,148 @@
+%   \c_@@_mathml_write_after_tl,\c_@@_mathml_write_final_tl} 
+%    \begin{macrocode}
+\tl_const:Nn \c_@@_mathml_write_init_tl 
+  {
+    <!DOCTYPE~html>
+    \iow_newline:
+    <html>
+    \iow_newline:
+  }
+\tl_new:N \l_@@_mathml_write_before_tl  
+\tl_const:Nn \c_@@_mathml_write_after_tl 
+  {
+    \iow_newline:
+    </div>
+    \iow_newline:
+  }  
+\tl_const:Nn \c_@@_mathml_write_final_tl 
+  {
+    </html>
+  }  
+%    \end{macrocode}
+% \end{macro}
 
+% \begin{socketdecl}{tagsupport/math/mathml/write/prepare}
+% To prepare the hash and the starting command we use a socket, so 
+% that both the dummy and luamml can make use of it.
+%    \begin{macrocode}
+\socket_new:nn {tagsupport/math/mathml/write/prepare}{0}
+%    \end{macrocode}
+% \end{socketdecl}
+%
+% \begin{plugdecl}{On}
+%    \begin{macrocode}
+\socket_new_plug:nnn{tagsupport/math/mathml/write/prepare}{On}
+  {
+    \str_set:NV\l_@@_tmpa_str\l_@@_content_AF_source_tl
+    \str_replace_all:Nnn\l_@@_tmpa_str{&}{&amp;}
+    \str_replace_all:Nnn\l_@@_tmpa_str{<}{&lt;}
+    \tl_set:Nn \l_@@_mathml_write_before_tl 
+      {
+        <div>
+        \iow_newline:
+        <h2>\c_backslash_str mml\c_space_tl \int_use:N \g_@@_math_total_int </h2>
+        \iow_newline:
+        <p>\l_@@_tmpa_str</p>
+        \iow_newline:
+        <p>\l_@@_content_hash_tl </p>
+        \iow_newline:
+      }   
+  }
+%    \end{macrocode}
+% \end{plugdecl}
 %
+% With luatex we automatically generate mathml with luamml if the package
+% can be loaded and unicode-math is detected. 
+% We start the process in the begindocument/end hook
+% so that the reading from a previous compilation can happen before!
+%    \begin{macrocode}
+\sys_if_engine_luatex:T
+ {
+   \file_if_exist:nT { luamml.sty }
+     {
+       \RequirePackage  { luamml }
+       \AddToHook{begindocument/end}
+         { 
+           \bool_if:NT \g__tag_math_luamml_bool 
+             { 
+               \@ifpackageloaded { unicode-math }
+                 { \@@_luamml_activate_write: }
+                 { \msg_warning:nn { tag }{ unicode-math-missing } }               
+             } 
+         }
+     }
+ }
+\msg_new:nnn { tag }{ unicode-math-missing } 
+  {
+    The~package~unicode-math~is~missing\\
+    luamml~will~not~create~an~MathML~file.\\
+    To~avoid~this~warning~load~unicode-math~or~disable~luamml\\
+    with~\tl_to_str:n{\tagpdfsetup{math/mathml/luamml=false}}
+  } 
+\cs_new_protected:Npn \@@_luamml_activate_write:
+ {
+%    \end{macrocode}
+% to avoid that nothing is written in the first run, we must activate the sockets:
+%    \begin{macrocode}
+   \bool_gset_true:N\g__tag_math_mathml_AF_bool
+   \AssignSocketPlug{tagsupport/math/struct/begin}{test-mathml}
+   \AssignSocketPlug{tagsupport/math/struct/end}{test-mathml}
+   \AssignSocketPlug{tagsupport/math/substruct/begin}{single}
+   \AssignSocketPlug{tagsupport/math/substruct/end}{single} 
+   \int_set:Nn \l__luamml_pretty_int { 5 }
```

Comment:
The pretty printing is a bit spacy and not consistent with mathml-core which doesn't trim white space so

```
<mi>
x
</mi>
```

would be better as

```
<mi>x</mi>
```

and only add pretty-print newlines between elements not in text content.  But this is a comment for luamml and shuould not affect this merge

Reply:
Do we want to adhere to mathml-core???

Reply:
> Do we want to adhere to mathml-core???

I think in the end we will need an option to configure luamml to target core.

As far as I can tell, the PDF specifications are largely silent about what mathml may be used either in associated files or if using the direct mathml structure tagging. So in practice it probably depends on what the screen readers implement.

If you target full mathml3 (including the "elementary math" elements for displayed sums and products, long division etc), then current generation browsers can display it, but using relatively large amounts of javascript to ployfill in the needed support. It is, to say the least, unclear how much of that would be accessible to a screen reader using mathml embedded in a pdf.  mathml-core is (designed to be) directly displayed in a browser requiring no javascript support and just requirng an opentype math font. So in practice is likely to be the base level of support targeted by screen readers and so probably the base level that should be targeted by systems generating mathml associated files to be used with an unknown reading application.

Reply:
"As far as I can tell, the PDF specifications are largely silent about what mathml may be used either in associated files or if using the direct mathml structure tagging."

Well, on the last part: there is a url reference to some particular MathML namespace, but that can easily be amended.

There is no reason to restrict the content of an associated file; but it is probably useful, and maybe compulsory, to sate what type of file it is, at some level of formality.

Reply:
@car222222 

> Well, on the last part: there is a url reference to some particular MathML namespace, but that can easily be amended.

All versions of MathML from 1.0 to 4.0 and mathml-core all use the same mathml namespace, so that namespace URI is just a slightly more formal way of saying "it's mathml" 

> There is no reason to restrict the content of an associated file;
If the reason for associating the  file is to improve accessibility then you need to use some format that AT systems can use. That is currently largely unspecified I think. In practice the only current pdf readers that can use the MathML AF in a pdf and do anything with it ultimately rely on mathcat to generate the speech which means that a largeish subset of mathml3 presentation is supported along with the intent attribute (which is a draft mathml4 feature being added specifically to improve accessibility and generated speech text in particular)